core/src/backend/video.rs

```diff
@@ -0,0 +1,159 @@
+//! Video decoder backends
+
+use crate::backend::render::{BitmapInfo, RenderBackend};
+use generational_arena::{Arena, Index};
+use swf::{VideoCodec, VideoDeblocking};
+
+mod software;
+
+pub use crate::backend::video::software::SoftwareVideoBackend;
+
+pub type VideoStreamHandle = Index;
+
+pub type Error = Box<dyn std::error::Error>;
+
+/// An encoded video frame of some video codec.
+pub struct EncodedFrame<'a> {
+    /// The codec used to encode the frame.
+    pub codec: VideoCodec,
+
+    /// The raw bitstream data to funnel into the codec.
+    pub data: &'a [u8],
+
+    /// A caller-specified frame ID. Frame IDs must be consistent between
+    /// subsequent uses of the same data stream.
+    pub frame_id: u32,
+}
+
+impl<'a> EncodedFrame<'a> {
+    /// Borrow this frame's data.
+    pub fn data(&'a self) -> &'a [u8] {
+        &self.data
+    }
+}
+
+/// What dependencies a given video frame has on any previous frames.
+pub enum FrameDependency {
+    /// This frame has no reference frames and can be seeked to at any time.
+    Keyframe,
+
+    /// This frame has one reference frame which must be completely decoded
+    /// before this frame can be decoded.
+    ///
+    /// The reference frame is the previous frame in the video stream.
+    LastFrame,
+}
```

Comment:
Is this an `enum` instead of a simple `is_keyframe : bool` for purely descriptive purposes (and good coding practice), or because it is fully expected that backwards-, bidirectional-, or non-consequential dependencies will have to be described? If the latter, why not add them now as options? Sure, they can always be added later, but I hope my point still makes sense.

Reply:
Both.

In general, `bool`-valued parameters hide function - if you have a function that takes multiple `bool`s, you don't know what each of them do. Enumerating your `bool` parameters makes the API more self-descriptive and should be done except in cases where the meaning of the parameter is obvious (or there is no meaning for whatever reason). Furthermore, an enumeration can be extended if you later need more than two values - which we might.

More advanced codecs have more complicated frame dependencies. Some of them don't need to be in this enum - VP6 golden frames are bounded by the local GOP (group-of-pictures), so their presence does not impact the "can I seek to this frame" question that this enum answers. They do stretch the definition of the `LastFrame` value, though. H.264 has another feature, however: B-Frames, which need a future frame to be decoded first for reference. So we would need a third enum value to signal future dependencies - but I don't know nearly enough about H.264 to know what the seek semantics are of B-Frames, how many future frames you need to pump into the decoder, and so on.

Looking at all of this, I'm starting to think these enum variants should be named `None`, `Past`, and (presumably) `Future` or `All`. I'm not going to make the change unless you're aware of it and willing to handle editing your VP6 branch to handle that, though.

Reply:
Yeah, falling into the "boolean trap" can be unfortunate indeed.
Feel free to make any changes you wish, of course; I'll follow through with them on my branch upon rebasing.
The VP6 branch is still a WIP anyways (I was _just_ going to add the keyframe "detection" right about now).

Reply:
I've done this. You now have `FrameDependency::None` and `Past`.